### PR TITLE
feat: Add StepfunctionsStack for agent evaluation workflow

### PR DESCRIPTION
**Pull request by ifeojo**
*Issue #, if available:*

*Description of changes:*
feat: Add StepfunctionsStack for agent evaluation workflow

- Implement StepfunctionsStack with AWS CDK
- Add Lambda functions for various steps in the workflow:
  - generate_map
  - check_agent_status
  - create_alias
  - delete_alias
  - run_test
  - update_bedrock_agent
- Create Step Functions state machine for agent evaluation process
- Set up EventBridge rule to trigger workflow on S3 object creation
- Configure necessary IAM roles and permissions
- Implement error handling and retry logic in the state machine

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by sharonxiaohanli — CHANGES_REQUESTED** [review]
Thank you for submitting the PR. Please add gitignore to ignore the cached files `__pycache__`.  We have code lint check in the CICD pipeline as well https://github.com/awslabs/agent-evaluation/blob/main/.github/workflows/ci.yml. Please make sure you run those commands locally and fix those formatting issues before submitting the PR.

File: stepfunctions/stepfunctions/functions/check_agent_status_1/.~c9_invoke_Zi2ZN1.py
Comment (sharonxiaohanli) [inline stepfunctions/stepfunctions/functions/check_agent_status_1/.~c9_invoke_Zi2ZN1.py:1]: what's this file for?

File: stepfunctions/app.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+    # Uncomment the next line to specialize this stack for the AWS Account
+    # and Region that are implied by the current CLI configuration.
+
+    #env=cdk.Environment(account=os.getenv('CDK_DEFAULT_ACCOUNT'), region=os.getenv('CDK_DEFAULT_REGION')),
+
+    # Uncomment the next line if you know exactly what Account and Region you
+    # want to deploy the stack to. */
+
+    env=cdk.Environment(region='us-east-1'),
```
Comment (sharonxiaohanli) [inline stepfunctions/app.py:23]: is there a reason to hard code the region?

**Review by sharonxiaohanli — CHANGES_REQUESTED** [review]

File: stepfunctions/stepfunctions/agenteval/__init__.py
Comment (sharonxiaohanli) [inline stepfunctions/stepfunctions/agenteval/__init__.py:1]: same here. why the agenteval core code is copied over? It should be used as an layer.

File: stepfunctions/stepfunctions/functions/check_agent_status_1/index.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
+import boto3
+import json
+import os
+
+s3_client = boto3.client('s3')
+bedrock_agent = boto3.client('bedrock-agent')
+
+# from aws_lambda_powertools import Logger, Tracer
```
Comment (sharonxiaohanli) [inline stepfunctions/stepfunctions/functions/check_agent_status_1/index.py:8]: For clean code, please remove those commented-out lines. But I think it's actually helpful to have the Logger from the powertool. And you can log `response` for any troubleshooting :). Not a blocking comment. I'll leave it to you to decide on the logging.

File: stepfunctions/stepfunctions/functions/check_agent_status_2/index.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
+import boto3
+import json
+import os
+
+s3_client = boto3.client('s3')
+bedrock_agent = boto3.client('bedrock-agent')
+
+# from aws_lambda_powertools import Logger, Tracer
```
Comment (sharonxiaohanli) [inline stepfunctions/stepfunctions/functions/check_agent_status_2/index.py:8]: same as above

File: stepfunctions/stepfunctions/functions/create_alias/index.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    
+    agent_alias = str(uuid.uuid4())
+    agent_id = event["update_output"]["agentid"]
+    
+    alias_resp = bedrock_agent.create_agent_alias(
+    agentAliasName=agent_alias,
+    agentId=agent_id
+    )
+    
+    print(alias_resp)
```
Comment (sharonxiaohanli) [inline stepfunctions/stepfunctions/functions/create_alias/index.py:18]: logger from the powertool produces cleaner logging format. If you decide to use powertools across all lambda functions, please make it consistent :)

File: stepfunctions/stepfunctions/functions/delete_alias/index.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    # TODO implement
+
+    #pass in from step function but for now
+    
+    agent_id = event["agent_id"]
+    agent_alias_id = event["agent_alias_id"]
+    
+    bedrock_agent = boto3.client('bedrock-agent')
+    
+    response = bedrock_agent.delete_agent_alias(
```
Comment (sharonxiaohanli) [inline stepfunctions/stepfunctions/functions/delete_alias/index.py:16]: check the status of `response` and log error if needed before sending back `200`.

File: stepfunctions/stepfunctions/functions/delete_alias/index.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
+import json
+import boto3
+import uuid
+import os
+
+def handler(event, context):
+    # TODO implement
+
+    #pass in from step function but for now
```
Comment (sharonxiaohanli) [inline stepfunctions/stepfunctions/functions/delete_alias/index.py:9]: Can you elaborate in the comment about the TODO here?

File: stepfunctions/stepfunctions/functions/generate_map/index.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
+import boto3
+import json
+import os
+
+s3_client = boto3.client('s3')
+
+# from aws_lambda_powertools import Logger, Tracer
```
Comment (sharonxiaohanli) [inline stepfunctions/stepfunctions/functions/generate_map/index.py:7]: delete commented out lines or use logger from powertools to log

File: stepfunctions/stepfunctions/functions/run_test/index.py [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+    yaml_data = {
+        'evaluator': {
+            'model': 'claude-3'
+        },
+        'target': {
+            'type': 'bedrock-agent',
+            'bedrock_agent_id': agent_id,
+            'bedrock_agent_alias_id': agent_alias_id
+        },
+        # 'user_profile':{
```
Comment (sharonxiaohanli) [inline stepfunctions/stepfunctions/functions/run_test/index.py:47]: delete the commented out lines.

File: stepfunctions/stepfunctions/functions/run_test/index.py [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+        
+    return{
+        'created_at': created_at,
+        'finished_at':finished_at,
+        'target_type': yaml_data["target"]["type"],
+        'status': status,
+        'test_passed_rate':test_passed_rate
+    }
+        
+    # finally:
```
Comment (sharonxiaohanli) [inline stepfunctions/stepfunctions/functions/run_test/index.py:150]: delete commented out block.

Comment (sharonxiaohanli) [inline stepfunctions/stepfunctions/functions/run_test/index.py:150]: delete commented out block.

File: stepfunctions/stepfunctions/functions/update_bedrock_agent/index.py [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+def handler(event, context):
+    # TODO implement
+
+    #pass in from step function but for now
+    
+    agent_id = event["agent_id"]
+    agent_name=event["agent_name"]
+    agent_role = os.environ['AGENT_ROLE']
+
+    # role = "arn:aws:iam::905418302891:role/service-role/AmazonBedrockExecutionRoleForAgents_LED91O3XKK"
```
Comment (sharonxiaohanli) [inline stepfunctions/stepfunctions/functions/update_bedrock_agent/index.py:15]: This line MUST be deleted :D .

File: stepfunctions/stepfunctions/functions/update_bedrock_agent/index.py [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+        agentId=agent_id,
+        agentName=agent_name,
+        agentResourceRoleArn=agent_role,
+        foundationModel=model,
+        instruction=instruction,
+        
+    )
+    
+    prep_resp = bedrock_agent.prepare_agent(agentId=agent_id)
+
```
Comment (sharonxiaohanli) [inline stepfunctions/stepfunctions/functions/update_bedrock_agent/index.py:31]: better to log out the response and add try/catch handler block.

File: stepfunctions/stepfunctions/stepfunctions_stack.py [hunk 12]
```diff
@@ -0,0 +1,432 @@
...
+        )
+        
+        update_agent_function.add_to_role_policy(
+            iam.PolicyStatement(
+                actions=["bedrock:*","iam:PassRole","iam:ListRoles"],
+                resources=["*"],
+            )  
+        )
+
+        # The code that defines your stack goes here
```
Comment (sharonxiaohanli) [inline stepfunctions/stepfunctions/stepfunctions_stack.py:426]: remove the commented out block

File: stepfunctions/stepfunctions/stepfunctions_stack.py [hunk 13]
```diff
@@ -0,0 +1,432 @@
...
+        sfn.Condition.string_equals("$.status_output_1.agentstatus", "VERSIONING")
+        )
+        
+        wait_step= sfn.Wait(
+            self,
+            "Wait1",
+            time=sfn.WaitTime.duration(Duration.seconds(30))
+            )
+            
+        # first_choice_def = first_choice.when(condition1, wait_step.next(get_status_step_1).next(first_choice)).otherwise(update_agent_step).afterwards()
```
Comment (sharonxiaohanli) [inline stepfunctions/stepfunctions/stepfunctions_stack.py:151]: remove this line

File: stepfunctions/tests/unit/test_stepfunctions_stack.py [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+from stepfunctions.stepfunctions_stack import StepfunctionsStack
+
+# example tests. To run these tests, uncomment this file along with the example
+# resource in stepfunctions/stepfunctions_stack.py
+def test_sqs_queue_created():
+    app = core.App()
+    stack = StepfunctionsStack(app, "stepfunctions")
+    template = assertions.Template.from_stack(stack)
+
+#     template.has_resource_properties("AWS::SQS::Queue", {
```
Comment (sharonxiaohanli) [inline stepfunctions/tests/unit/test_stepfunctions_stack.py:13]: remove the block.

File: stepfunctions/stepfunctions/functions/update_bedrock_agent/index.py (hunk 10 above)
Comment (ifeojo) [reply stepfunctions/stepfunctions/functions/update_bedrock_agent/index.py:15]: oops!!! :D

**Review by sharonxiaohanli — APPROVED** [review]
LGTM! Thank you for working on it

**Pull request merged**